examples/hpc/dla-driver/src/lib.rs

```diff
@@ -18,291 +106,501 @@ macro_rules! set_bits {
     };
 }
 
-pub fn dla_write_str(s: &str) {
-    for b in s.as_bytes() {
-        unsafe { ptr::write_volatile(DLA0_ADDR as *mut u8, *b) };
-    }
-}
-pub fn dla_write(offset: usize, value: u8) {
-    unsafe { ptr::write_volatile((offset) as *mut u8, value) };
+macro_rules! get_bits {
+    ($reg:expr, $mask:expr) => {
+        ($reg & ($mask as u32)) as u32
+    };
 }
 
-pub fn dla_write_reg(offset: usize, value: u32) {
-    unsafe { ptr::write_volatile((DLA0_ADDR + offset) as *mut u32, value) }
-}
+impl Dla {
+    pub fn new() -> Self {
+        Dla {
+        }
+    }
+    pub fn write_str(&self, s: &str) {
+        for b in s.as_bytes() {
+            unsafe { ptr::write_volatile(DLA0_ADDR as *mut u8, *b) };
+        }
+    }
+    pub fn write(&self, offset: usize, value: u8) {
+        unsafe { ptr::write_volatile((offset) as *mut u8, value) };
+    }
 
-pub fn dla_read_reg(offset: usize) -> u32 {
-    unsafe { ptr::read_volatile((DLA0_ADDR + offset) as *mut u32) }
-}
+    pub fn write_u32(&self, offset: usize, value: u32) {
+        unsafe { ptr::write_volatile((DLA0_ADDR + offset) as *mut u32, value) }
+    }
 
-pub fn dla_read(buf: &mut [u8], len: usize, offset: usize) {
-    for i in 0..len {
-        unsafe { buf[i] = ptr::read_volatile((DLA0_ADDR + offset + i) as *mut u8) }
+    pub fn read_u32(&self, offset: usize) -> u32 {
+        unsafe { ptr::read_volatile((DLA0_ADDR + offset) as *mut u32) }
     }
-}
 
-pub fn dla_write_data_bank(offset: usize, buf: &mut [u8]) {
-    sprint!("\nWrite to bank {:#x}, data: {:?}", offset, buf);
-    for (i, b) in buf.iter().enumerate() {
-        unsafe { ptr::write_volatile((MEMORY_BANK_BASE_ADDR + offset + i) as *mut u8, *b) };
+    pub fn read(&self, buf: &mut [u8], len: usize, offset: usize) {
```

Comment:
Recommended parameter ordering: `offset, len, &mut buf`

Reply:
Recommended fn name: `read_bytes`

Reply:
Renamed function and reordered parameters as suggested.